### PR TITLE
DAOS-4661 Test: Adding test case and Python APi for pool target query.

### PR DESCRIPTION
**Pull request by ravalsam**
- Adding python API support for target_query.
- Adding Avocado test to validate the specific object data is
  writing to the expected number of targets

Test-tag: iorsmall pool_query_target

Signed-off-by: Samir Raval <samir.raval@intel.com>

### CODE REVIEW

File: src/tests/ftest/pool/target_query.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+        # Get the size after writing the IOR data
+        latest_nvme_size = self.target_space_info()
+        self.log.info("Latest Target space = %s", latest_nvme_size)
+
+        # Verify the target sizes is only increasing based on object layout
+        # specified in yaml file
+        target_count = 0
+        for count, _ in enumerate(initial_nvme_size):
+            if latest_nvme_size[count] != initial_nvme_size[count]:
```
Comment (daltonbohning) [inline src/tests/ftest/pool/target_query.py:71]: Suggestion: I'd recommend either
```suggestion
        for count, initial_size in enumerate(initial_nvme_size):
            if latest_nvme_size[count] != initial_size:
```
Or
```suggestion
        for count in range(len((initial_nvme_size)):
            if latest_nvme_size[count] != initial_nvme_size[count]:
```
Though pylint might complain about the latter.

File: src/tests/ftest/pool/target_query.yaml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+container:
+    type: POSIX
+    control_method: daos
+ior:
+  api: "DFS"
+  client_processes:
+   np: 1
+  dfs_destroy: False
+  iorflags:
+   flags: "-w -F -k -G 1 -vv"
```
Comment (daltonbohning) [inline src/tests/ftest/pool/target_query.yaml:58]: Minor spacing - This will eventually be enforced by yaml linting
```suggestion
  client_processes:
    np: 1
  dfs_destroy: False
  iorflags:
    flags: "-w -F -k -G 1 -vv"
```

File: src/tests/ftest/pool/target_query.yaml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  test_file: /testFile
+  repetitions: 1
```
Comment (daltonbohning) [inline src/tests/ftest/pool/target_query.yaml:60]: Recommend removing as `repetitions: 1` is the default

File: src/tests/ftest/pool/target_query.yaml (hunk 2 above)
Comment (ravalsam) [reply src/tests/ftest/pool/target_query.yaml:58]: Done

File: src/tests/ftest/pool/target_query.yaml (hunk 3 above)
Comment (ravalsam) [reply src/tests/ftest/pool/target_query.yaml:60]: Done

File: src/tests/ftest/pool/target_query.yaml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    - server-B
+    - server-C
+    - server-D
+    - server-E
+  test_clients:
+    - client-A
+timeout: 800
+setup:
+  start_agents_once: False
+  start_servers_once: False
```
Comment (daltonbohning) [inline src/tests/ftest/pool/target_query.yaml:13]: Why is this necessary? Is it not enough for each mux variant to just create a new pool?

**Review by daltonbohning — DISMISSED** [review]
LGTM, but one question

File: src/tests/ftest/pool/target_query.yaml (hunk 4 above)
Comment (ravalsam) [reply src/tests/ftest/pool/target_query.yaml:13]: Yes its copy paste. let me remove it

**Review by daltonbohning — DISMISSED** [review]

**Review by saurabhtandan — DISMISSED** [review]
LGTM

**Review by phender — COMMENTED** [review]
Some questions and possible improvements.  Probably not enough for a -1, depending on the answers.

File: src/tests/ftest/pool/target_query.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        """
+        nvme_rank_space = []
+        self.pool.connect()
+
+        # Collect the space information for all the ranks and targets
+        for _rank in range(self.server_count):
+            for _target in range(self.server_managers[0].get_config_value("targets")):
+                result = self.pool.pool.target_query(_target, _rank)
+                nvme_rank_space.append(result.ta_space.s_free[1])
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/pool/target_query.py:37]: There are two filed under Daos_Space. Free and Total. The Free space is the one which will change based on object type it's written. Total space is not going to have any impact and the scope of the ticket was to only verify the free space after writing and make sure it only changes for the required targets.

File: src/tests/ftest/pool/target_query.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                  on object type.
+
+        :avocado: tags=all,full_regression
+        :avocado: tags=hw,large
+        :avocado: tags=pool
+        :avocado: tags=pool_target_query
+        """
+        self.update_ior_cmd_with_pool()
+        # Check the initial size of all targets
+        initial_nvme_size = self.target_space_info()
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/pool/target_query.py:57]: Yes. I will add it.

File: src/tests/ftest/pool/target_query.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        # Verify the target sizes is only increasing based on object layout
+        # specified in yaml file
+        target_count = 0
+        for count, initial_size in enumerate(initial_nvme_size):
+            if latest_nvme_size[count] != initial_size:
+                target_count += 1
+        if target_count != self.target_usage_count:
+            self.fail("Expected to change the free space capacity for {} targets,"
+                      " But it's changed for {} targets".format(self.target_usage_count,
+                                                               str(target_count)))
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/pool/target_query.py:76]: I will update it.

File: src/tests/ftest/pool/target_query.yaml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    flags: "-w -F -k -G 1 -vv"
+  test_file: /testFile
+  transfer_size: 1M
+  block_size: 1G
+  objectclass: !mux
+    #SX Should write to all Targets
+    SX:
+      dfs_oclass: "SX"
+      target_usage_count: 20
+    #RP_2G1 is Group G1 so will write to Two replica targets with in 1 groups only.
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/pool/target_query.yaml:64]: Sure, I can add it.

File: src/tests/ftest/pool/target_query.yaml [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      dfs_oclass: "EC_4P2GX"
+      target_usage_count: 18
+      notes: "EC_4P2GX is Group GX with EC_4P2, so it should write to 4data+2parity (6 targets),
+              among max groups (In this case 3 so 6x3=18 targets)"
+    EC_8P2GX:
+      dfs_oclass: "EC_8P2GX"
+      target_usage_count: 20
+      notes: "EC_8P2GX is Group GX with EC_8P2, so it should write to 8data+2parity (10 targets),
+              among max groups (In this case 2 so 10x2 targets)"
+
```
Comment (ashleypittman) [inline src/tests/ftest/pool/target_query.yaml:87]: This is a new check but will need to be fixed.
```suggestion
              among max groups (In this case 2 so 10x2 targets)"
```

File: src/tests/ftest/pool/target_query.yaml [hunk 10]
```diff
@@ -57,30 +57,30 @@ ior:
   transfer_size: 1M
   block_size: 1G
   objectclass: !mux
-    #SX Should write to all Targets
     SX:
       dfs_oclass: "SX"
       target_usage_count: 20
-    #RP_2G1 is Group G1 so will write to Two replica targets with in 1 groups only.
+      notes: "SX should write to all Targets"
```
Comment (daltonbohning) [inline src/tests/ftest/pool/target_query.yaml:63]: I like the idea of putting notes/description in the yaml... :)

**Review by daltonbohning — APPROVED** [review]

**Review by saurabhtandan — APPROVED** [review]
LGTM
